File: src/main/java/li/cil/oc/api/driver/UpgradeContainer.java
package li.cil.oc.api.driver;

import net.minecraft.item.ItemStack;

/**
 * May be implemented by drivers for robot upgrades that act as hotswap bays,
 * i.e. which can be installed into the 'dynamic' slots, and provide on-the-fly
 * changeable upgrade slots (i.e. which can be changed in the robot GUI,
 * without disassembling and re-assembling the robot).
 * <p/>
 * These drivers will not be queried for environments. The reported tier is the
 * maximum tier supported in the dynamic slot they provide.
 */
public interface UpgradeContainer extends Item {
    /**
     * The type of slot provided as the dynamic slot. This will usually be
     * for other upgrades, but may be for any type of item component.
     * <p/>
     * While the driver's own type implicitly has to be 'Upgrade' and could
     * therefore be used instead, this makes the intention more clear.
     *
     * @param stack the item stack to get the provided slot type for.
     * @return the slot type provided by that dynamic slot upgrade.
     */
    Slot providedSlot(ItemStack stack);

    /**
     * The maximum item tier of the items that can be placed into the slot
     * provided by the specified container.
     * <p/>
     * This will usually be equal to the container's tier.
     *
     * @param stack the item stack to the the supported tier for.
     * @return the maximum tier supported by that dynamic slot upgrade.
     */
    int providedTier(ItemStack stack);
}

File: src/main/scala/li/cil/oc/server/driver/item/UpgradeContainer.scala
package li.cil.oc.server.driver.item

import li.cil.oc.api.driver.Container
import li.cil.oc.api.driver.EnvironmentHost
import li.cil.oc.common.Slot
import li.cil.oc.common.item
import li.cil.oc.Items
import li.cil.oc.api
import net.minecraft.item.ItemStack

object UpgradeContainer extends Item with Container {
  override def worksWith(stack: ItemStack) =
    isOneOf(stack, api.Items.get("upgradeContainer1"), api.Items.get("upgradeContainer2"), api.Items.get("upgradeContainer3"))

  override def createEnvironment(stack: ItemStack, host: EnvironmentHost) = null

  override def slot(stack: ItemStack) = Slot.Container

  override def providedSlot(stack: ItemStack) = Slot.Upgrade

  override def providedTier(stack: ItemStack) = tier(stack)

  override def tier(stack: ItemStack) =
    Items.multi.subItem(stack) match {
      case Some(container: item.UpgradeContainerUpgrade) => container.tier
      case _ => 0
    }
}
